# Telecom Churn case study using PCA and Ensemble

### Problem Statement
In the telecom industry, customers are able to choose from multiple service providers and actively switch from one operator to another. In this highly competitive market, the telecommunications industry experiences an average of 15-25% annual churn rate. Given the fact that it costs 5-10 times more to acquire a new customer than to retain an existing one, customer retention has now become even more important than customer acquisition.

For many incumbent operators, retaining high profitable customers is the number one business
goal. To reduce customer churn, telecom companies need to predict which customers are at high risk of churn. In this project, you will analyze customer-level data of a leading telecom firm, build predictive models to identify customers at high risk of churn.

In this competition, your goal is to build a machine learning model that is able to predict churning customers based on the features provided for their usage.

### Goal
It is your job to predict if a customer will churn, given the ~170 columns containing customer behavior, usage patterns, payment patterns, and other features that might be relevant. Your target variable is "churn_probability"
Note: Make sure your accuracy is greater than the sample submission that is present in the leaderboard

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

from sklearn.decomposition import IncrementalPCA
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.preprocessing import MinMaxScaler

## Step 1. Reading, Understanding and Visualizing the data

In [2]:
telecom_train_df = pd.read_csv('train (1).csv')
telecom_test_df = pd.read_csv('test (1).csv')

In [3]:
telecom_train_df.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0   0        109             0.0             0.0             0.0   
1   1        109             0.0             0.0             0.0   
2   2        109             0.0             0.0             0.0   
3   3        109             0.0             0.0             0.0   
4   4        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   31.277   
1            6/30/2014            7/31/2014            8/31/2014    0.000   
2            6/30/2014            7/31/2014            8/31/2014   60.806   
3            6/30/2014            7/31/2014            8/31/2014  156.362   
4            6/30/2014            7/31/2014            8/31/2014  240.708   

    arpu_7  ...  sachet_3g_7  sachet_3g_8  fb_user_6  fb_user_7  fb_user_8  \
0   87.009  ...            0            0        NaN        NaN        NaN   
1  122.787  ...            0            0        NaN        1.0        NaN   
2  103.176  ...            0            0        NaN        NaN        NaN   
3  205.260  ...            0            0        NaN        NaN        NaN   
4  128.191  ...            1            0        1.0        1.0        1.0   

    aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  churn_probability  
0  1958         0.0         0.0         0.0                  0  
1   710         0.0         0.0         0.0                  0  
2   882         0.0         0.0         0.0                  0  
3   982         0.0         0.0         0.0                  0  
4   647         0.0         0.0         0.0                  0  

[5 rows x 172 columns]

In [4]:
telecom_test_df.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0  69999        109             0.0             0.0             0.0   
1  70000        109             0.0             0.0             0.0   
2  70001        109             0.0             0.0             0.0   
3  70002        109             0.0             0.0             0.0   
4  70003        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   91.882   
1            6/30/2014            7/31/2014            8/31/2014  414.168   
2            6/30/2014            7/31/2014            8/31/2014  329.844   
3            6/30/2014            7/31/2014            8/31/2014   43.550   
4            6/30/2014            7/31/2014            8/31/2014  306.854   

    arpu_7  ...  sachet_3g_6  sachet_3g_7  sachet_3g_8  fb_user_6  fb_user_7  \
0   65.330  ...            0            0            0        NaN        NaN   
1  515.568  ...            0            0            0        NaN        NaN   
2  434.884  ...            0            0            0        NaN        NaN   
3  171.390  ...            0            0            0        NaN        NaN   
4  406.289  ...            0            0            0        NaN        NaN   

   fb_user_8   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  
0        NaN  1692        0.00        0.00        0.00  
1        NaN  2533        0.00        0.00        0.00  
2        NaN   277      525.61      758.41      241.84  
3        NaN  1244        0.00        0.00        0.00  
4        NaN   462        0.00        0.00        0.00  

[5 rows x 171 columns]

In [5]:
telecom_train_df.shape

(69999, 172)

In [6]:
telecom_test_df.shape

(30000, 171)

In [7]:
telecom_train_df.columns

Index(['id', 'circle_id', 'loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou',
       'last_date_of_month_6', 'last_date_of_month_7', 'last_date_of_month_8',
       'arpu_6', 'arpu_7',
       ...
       'sachet_3g_7', 'sachet_3g_8', 'fb_user_6', 'fb_user_7', 'fb_user_8',
       'aon', 'aug_vbc_3g', 'jul_vbc_3g', 'jun_vbc_3g', 'churn_probability'],
      dtype='object', length=172)

In [8]:
telecom_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 172 entries, id to churn_probability
dtypes: float64(135), int64(28), object(9)
memory usage: 91.9+ MB


In [9]:
telecom_train_df.describe()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count  69999.000000    69999.0         69297.0         69297.0   
mean   34999.000000      109.0             0.0             0.0   
std    20207.115084        0.0             0.0             0.0   
min        0.000000      109.0             0.0             0.0   
25%    17499.500000      109.0             0.0             0.0   
50%    34999.000000      109.0             0.0             0.0   
75%    52498.500000      109.0             0.0             0.0   
max    69998.000000      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         69297.0  69999.000000  69999.000000  69999.000000  67231.000000   
mean              0.0    283.134365    278.185912    278.858826    133.153275   
std               0.0    334.213918    344.366927    351.924315    299.963093   
min               0.0  -2258.709000  -1289.715000   -945.808000      0.000000   
25%               0.0     93.581000     86.714000     84.095000      7.410000   
50%               0.0    197.484000    191.588000    192.234000     34.110000   
75%               0.0    370.791000    365.369500    369.909000    119.390000   
max               0.0  27731.088000  35145.834000  33543.624000   7376.710000   

        onnet_mou_7  ...   sachet_3g_7   sachet_3g_8     fb_user_6  \
count  67312.000000  ...  69999.000000  69999.000000  17568.000000   
mean     133.894438  ...      0.081444      0.085487      0.916325   
std      311.277193  ...      0.634547      0.680035      0.276907   
min        0.000000  ...      0.000000      0.000000      0.000000   
25%        6.675000  ...      0.000000      0.000000      1.000000   
50%       32.280000  ...      0.000000      0.000000      1.000000   
75%      115.837500  ...      0.000000      0.000000      1.000000   
max     8157.780000  ...     33.000000     41.000000      1.000000   

          fb_user_7     fb_user_8           aon    aug_vbc_3g    jul_vbc_3g  \
count  17865.000000  18417.000000  69999.000000  69999.000000  69999.000000   
mean       0.909544      0.890319   1220.639709     68.108597     65.935830   
std        0.286842      0.312501    952.426321    269.328659    267.899034   
min        0.000000      0.000000    180.000000      0.000000      0.000000   
25%        1.000000      1.000000    468.000000      0.000000      0.000000   
50%        1.000000      1.000000    868.000000      0.000000      0.000000   
75%        1.000000      1.000000   1813.000000      0.000000      0.000000   
max        1.000000      1.000000   4337.000000  12916.220000   9165.600000   

        jun_vbc_3g  churn_probability  
count  69999.00000       69999.000000  
mean      60.07674           0.101887  
std      257.22681           0.302502  
min        0.00000           0.000000  
25%        0.00000           0.000000  
50%        0.00000           0.000000  
75%        0.00000           0.000000  
max    11166.21000           1.000000  

[8 rows x 163 columns]

In [10]:
telecom_train_df.isna().sum()

id                     0
circle_id              0
loc_og_t2o_mou       702
std_og_t2o_mou       702
loc_ic_t2o_mou       702
                    ... 
aon                    0
aug_vbc_3g             0
jul_vbc_3g             0
jun_vbc_3g             0
churn_probability      0
Length: 172, dtype: int64

In [11]:
telecom_train_df.dtypes

id                     int64
circle_id              int64
loc_og_t2o_mou       float64
std_og_t2o_mou       float64
loc_ic_t2o_mou       float64
                      ...   
aon                    int64
aug_vbc_3g           float64
jul_vbc_3g           float64
jun_vbc_3g           float64
churn_probability      int64
Length: 172, dtype: object

In [12]:
# Columns which we have to impute as Zero as thy should not be dropped based on missing values becuase they are important.
rech_cols_to_impute = [x for x in telecom_train_df.columns if 'rech' in x and 
                       'count' not in x and 'date' not in x and 'num' not in x]
telecom_train_df[rech_cols_to_impute] = telecom_train_df[rech_cols_to_impute].apply(lambda x: x.fillna(0))
telecom_test_df[rech_cols_to_impute] = telecom_test_df[rech_cols_to_impute].apply(lambda x: x.fillna(0))


### Dropping rows for both train and test datasets
 - Rows which are empty

In [13]:
telecom_train_df.dropna(axis=0, how='all', inplace=True)
telecom_test_df.dropna(axis=0, how='all', inplace=True)

In [14]:
print(telecom_train_df.shape)
print(telecom_test_df.shape)

(69999, 172)
(30000, 171)


### Dropping columns for both train and test datasets

    - Not needed columns like Id, last_date_of_month_6 etc.
    - Columns having more than 70% values as null

In [15]:
# For us dates doesn't matter as long as customer is doing a recharge.
cols_to_delete = ['id', 'last_date_of_month_6', 'last_date_of_month_7', 'last_date_of_month_8',
                  'date_of_last_rech_data_6', 'date_of_last_rech_data_7', 'date_of_last_rech_data_8',
                  'date_of_last_rech_6', 'date_of_last_rech_7', 'date_of_last_rech_8'
                 ]
telecom_train_df.drop(cols_to_delete, axis=1, inplace=True)
telecom_test_df.drop(cols_to_delete, axis=1, inplace=True)

In [16]:
# to check the % of null values
round((pd.isnull(telecom_train_df).sum()/len(telecom_train_df.index)),2)[pd.isnull(telecom_train_df).sum()/ len(telecom_train_df.index) > 0.70].sort_values()

arpu_3g_8           0.74
night_pck_user_8    0.74
night_pck_user_7    0.74
arpu_2g_8           0.74
arpu_2g_7           0.74
fb_user_7           0.74
arpu_3g_7           0.74
fb_user_8           0.74
count_rech_3g_8     0.74
count_rech_3g_7     0.74
count_rech_2g_8     0.74
count_rech_2g_7     0.74
arpu_2g_6           0.75
count_rech_3g_6     0.75
night_pck_user_6    0.75
fb_user_6           0.75
arpu_3g_6           0.75
count_rech_2g_6     0.75
dtype: float64

In [17]:
# count of missing data > 70% columns in train dataset
len(round((pd.isnull(telecom_train_df).sum()/len(telecom_train_df.index)),2)[pd.isnull(telecom_train_df).sum()/ len(telecom_train_df.index) > 0.70].sort_values())

18

In [18]:
# count of missing data > 70% columns in test dataset
len(round((pd.isnull(telecom_test_df).sum()/len(telecom_test_df.index)),2)[pd.isnull(telecom_test_df).sum()/ len(telecom_test_df.index) > 0.70].sort_values())

18

### As we have more than 70% data as null for 18 columns in train and test, we are dropping these columns from dataset.

In [19]:
cols_to_delete = ((pd.isnull(telecom_train_df).sum()/len(telecom_train_df.index))[pd.isnull(telecom_train_df).sum()/ len(telecom_train_df.index) > 0.70].sort_values()).keys()

In [20]:
cols_to_delete = cols_to_delete.to_list()
print(cols_to_delete)

['arpu_3g_8', 'night_pck_user_8', 'arpu_2g_8', 'count_rech_3g_8', 'fb_user_8', 'count_rech_2g_8', 'arpu_3g_7', 'fb_user_7', 'arpu_2g_7', 'night_pck_user_7', 'count_rech_2g_7', 'count_rech_3g_7', 'count_rech_3g_6', 'arpu_3g_6', 'arpu_2g_6', 'night_pck_user_6', 'fb_user_6', 'count_rech_2g_6']


In [21]:
telecom_train_df = telecom_train_df.drop(cols_to_delete, axis=1)

In [22]:
telecom_train_df.shape

(69999, 144)

In [23]:
telecom_test_df = telecom_test_df.drop(cols_to_delete, axis=1)

In [24]:
telecom_test_df.shape

(30000, 143)

## Lets check the data types of columns now

In [25]:
telecom_train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69999 entries, 0 to 69998
Columns: 144 entries, circle_id to churn_probability
dtypes: float64(117), int64(27)
memory usage: 77.4 MB


In [26]:
telecom_test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30000 entries, 0 to 29999
Columns: 143 entries, circle_id to jun_vbc_3g
dtypes: float64(117), int64(26)
memory usage: 33.0 MB


### We can see now we don't have any categorical data, all are numerical data only.

### Imputing the na values with
 - Median for numerical variables
 - Mode for categorical variables - Not needed in out this case.

In [27]:
# A generic method to impute na
def replace_na(df, columns, data_type):
    for col in columns:
        if data_type == 'categorical':
            value = pd.to_datetime(df[col]).mode()
        else:
            value = df[col].median()
        df[col].fillna(value, inplace=True)
    return df

In [28]:
telecom_train_df = replace_na(telecom_train_df, telecom_train_df.columns, 'numerical')
telecom_test_df = replace_na(telecom_test_df, telecom_test_df.columns, 'numerical')

### Lets check for unique entries in train and test dataset

In [29]:
def cols_to_delete_unique(df, df_columns, threshold=4):
    # threshold=4 means column has only 1 unique value
    # eg. telecom_train_df['circle_id'].describe().unique() --> array([30000.,   109.,     0.])
    # Here because there are all 109 in column, we have only 3 entries in array.
    cols_to_delete = []
    for column in df_columns:
        if len(df[column].describe().unique()) < threshold:
            cols_to_delete.append(column)
    
    return cols_to_delete

In [30]:
print(telecom_train_df.shape)
print(telecom_test_df.shape)

(69999, 144)
(30000, 143)


In [31]:
# on train data
telecom_train_df = telecom_train_df.drop(cols_to_delete_unique(telecom_train_df, telecom_train_df.columns), axis=1)
print(telecom_train_df.shape)

# on test data
telecom_test_df = telecom_test_df.drop(cols_to_delete_unique(telecom_test_df, telecom_test_df.columns), axis=1)
print(telecom_test_df.shape)

(69999, 134)
(30000, 133)


### Lets find the high value customers based on recharge done in 6th and 7th month and update dataframe

### For train data

In [32]:
# total data recharge in 6th, 7th and 8th months
telecom_train_df["total_data_recharge_amnt_6"] = telecom_train_df.total_rech_data_6 * telecom_train_df.av_rech_amt_data_6
telecom_train_df["total_data_recharge_amnt_7"] = telecom_train_df.total_rech_data_7 * telecom_train_df.av_rech_amt_data_7
telecom_train_df["total_data_recharge_amnt_8"] = telecom_train_df.total_rech_data_8 * telecom_train_df.av_rech_amt_data_8

# total amount spent on recharge in 6th, 7th and 8th months
telecom_train_df["total_recharge_amnt_6"] = telecom_train_df.total_rech_amt_6 + telecom_train_df.total_data_recharge_amnt_6
telecom_train_df["total_recharge_amnt_7"] = telecom_train_df.total_rech_amt_7 + telecom_train_df.total_data_recharge_amnt_7
telecom_train_df["total_recharge_amnt_8"] = telecom_train_df.total_rech_amt_8 + telecom_train_df.total_data_recharge_amnt_8

# average recharge for 6th and 7th month
telecom_train_df['average_amnt_6_7'] = (telecom_train_df["total_recharge_amnt_6"] + telecom_train_df["total_recharge_amnt_7"])/2

# 70th percentile of average_amnt_6_7
telecom_train_df['average_amnt_6_7'].quantile(.70)

477.5

## Filter dataset based on average_amnt_6_7 (70th percentile for train dataset is 477.5)

In [33]:
telecom_train_df = telecom_train_df[telecom_train_df["average_amnt_6_7"]>= telecom_train_df["average_amnt_6_7"].quantile(.70)]

In [34]:
telecom_train_df.shape

(21013, 141)

## Lets visualize the 8th month and average_amnt_6_7 months data to check behavior pattern of customers

In [35]:
# Look it later
#plt.scatter(telecom_train_df[['average_amnt_6_7', 'total_recharge_amnt_8']])
#plt.show()
#sns.
#telecom_train_df.plot.scatter(x='average_amnt_6_7', y='total_recharge_amnt_8')
#telecom_train_df.DataFrame(np.random.rand(10, 3), columns =['average_amnt_6_7', 'total_recharge_amnt_8'])

#telecom_train_df.plot.bar(x="average_amnt_6_7", y="total_recharge_amnt_8", rot=70);

#ig, ax = plt.subplots()

#ax.plot(telecom_train_df['average_amnt_6_7'], telecom_train_df['total_recharge_amnt_8'])
#plt.show()

### For test data

In [36]:
# total data recharge in 6th, 7th and 8th months
telecom_test_df["total_data_recharge_amnt_6"] = telecom_test_df.total_rech_data_6 * telecom_test_df.av_rech_amt_data_6
telecom_test_df["total_data_recharge_amnt_7"] = telecom_test_df.total_rech_data_7 * telecom_test_df.av_rech_amt_data_7
telecom_test_df["total_data_recharge_amnt_8"] = telecom_test_df.total_rech_data_8 * telecom_test_df.av_rech_amt_data_8

# total amount spent on recharge in 6th, 7th and 8th months
telecom_test_df["total_recharge_amnt_6"] = telecom_test_df.total_rech_amt_6 + telecom_test_df.total_data_recharge_amnt_6
telecom_test_df["total_recharge_amnt_7"] = telecom_test_df.total_rech_amt_7 + telecom_test_df.total_data_recharge_amnt_7
telecom_test_df["total_recharge_amnt_8"] = telecom_test_df.total_rech_amt_8 + telecom_test_df.total_data_recharge_amnt_8

# average recharge
telecom_test_df['average_amnt_6_7'] = (telecom_test_df["total_recharge_amnt_6"] + telecom_test_df["total_recharge_amnt_7"])/2

# 70th percentile of average_amnt_6_7
telecom_test_df['average_amnt_6_7'].quantile(.70)

478.0

### Filter dataset based on average_amnt_6_7 (70th percentile for test dataset is 478.0)

In [37]:
telecom_test_df = telecom_test_df[telecom_test_df["average_amnt_6_7"]>= telecom_test_df["average_amnt_6_7"].quantile(.70)]
telecom_test_df.shape

(9003, 140)

## Lets focus on Outliers now and treat them

In [38]:
# Generic method to remove outliers
def remove_outliers(df, features):
    for feature in features:
        q1 = df[feature].quantile(0.25)
        q3 = df[feature].quantile(0.99)
        iqr = q3-q1
        lower_value  = q1 - (1.5 * iqr)
        higer_value = q3 + (1.5 * iqr)
        df = df[(df[feature] <= higer_value) & (df[feature] >= lower_value)]
    return df

In [39]:
#telecom_train_df = remove_outliers(telecom_train_df, telecom_train_df.columns)
#telecom_test_df = remove_outliers(telecom_test_df, telecom_test_df.columns)

In [40]:
telecom_train_df.shape

(21013, 141)

In [41]:
telecom_test_df.shape

(9003, 140)

## Let's check for Data imbalance here

In [42]:
churn_data_rows = telecom_train_df['churn_probability'].sum()

In [43]:
churn_data_rows/len(telecom_train_df) * 100

8.275829248560415

### We can see that there is a data imbalance. We have only about 8.28% of data as Churn and 91.72% as not churn.
### So we need to apply data imbalance technique. We will use SMOTE.

## Dividing the data from telecom_train_df to train the model

In [44]:
y = telecom_train_df['churn_probability']
X = telecom_train_df.drop(['churn_probability'], axis=1)

telecom_train_df.drop('churn_probability', axis=1, inplace=True)

In [45]:
# Splitting the data into test and train
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, test_size=0.3, random_state=100)

## Rescaling of the variables

- We will use Min-Max scaling (Normalization) --> compresses all the data between 0 and 1

In [46]:
# Creating MinMaxScaler object
scaler = MinMaxScaler()
# fit on data
# removed churn_probability becuase test data is not having it.
# Also churn_probability is having 0 and 1 only. So it need not be scaled.
train_numerical_columns = telecom_train_df.columns.to_list()
X_train[train_numerical_columns] = scaler.fit_transform(X_train[train_numerical_columns])
X_train.head()
# So all numberic values are now between 0 and 1

arpu_6    arpu_7    arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  \
6968   0.127277  0.274576  0.081830     0.004624     0.004018     0.002985   
42203  0.156166  0.094305  0.083632     0.066057     0.041462     0.074009   
18406  0.151447  0.122733  0.087708     0.015153     0.016084     0.008306   
21455  0.171177  0.128305  0.077781     0.009636     0.010909     0.006579   
66143  0.138174  0.087866  0.074838     0.000000     0.006171     0.001633   

       offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  ...  \
6968       0.011537      0.040985      0.006554       0.000000  ...   
42203      0.065428      0.053213      0.049054       0.000000  ...   
18406      0.056327      0.259820      0.136117       0.005055  ...   
21455      0.081607      0.121132      0.038009       0.005166  ...   
66143      0.000000      0.033354      0.020053       0.000000  ...   

       aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  total_data_recharge_amnt_6  \
6968     0.000000    0.002051    0.000000                    0.000000   
42203    0.000000    0.000000    0.000000                    0.000000   
18406    0.000000    0.000000    0.000000                    0.000000   
21455    0.029661    0.043865    0.039021                    0.000000   
66143    0.098218    0.000000    0.000000                    0.043938   

       total_data_recharge_amnt_7  total_data_recharge_amnt_8  \
6968                     0.000000                    0.000000   
42203                    0.000000                    0.003787   
18406                    0.000000                    0.000000   
21455                    0.011038                    0.011124   
66143                    0.000000                    0.016759   

       total_recharge_amnt_6  total_recharge_amnt_7  total_recharge_amnt_8  \
6968                0.000000               0.096264               0.000000   
42203               0.009557               0.009933               0.016496   
18406               0.008632               0.019367               0.017273   
21455               0.015034               0.032992               0.023503   
66143               0.045971               0.006063               0.026593   

       average_amnt_6_7  
6968           0.069171  
42203          0.002008  
18406          0.009450  
21455          0.026840  
66143          0.029892  

[5 rows x 140 columns]

In [47]:
from imblearn.over_sampling import SMOTE

In [48]:
X_resampled, y_resampled = SMOTE().fit_resample(X_train, y_train)

In [49]:
X_resampled.shape

(27020, 140)

In [50]:
sum(y_resampled)/len(y_resampled) * 100

50.0

## Now there is no data imbalance.

In [51]:
# Lets fit the scaler for train data as well
test_numerical_columns = telecom_test_df.columns.to_list()
telecom_test_df[test_numerical_columns] = scaler.transform(telecom_test_df[test_numerical_columns])
telecom_test_df.head()

arpu_6    arpu_7    arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  \
1   0.150615  0.099685  0.064554     0.010236     0.005052     0.001845   
2   0.145864  0.095230  0.083592     0.001022     0.000963     0.000781   
21  0.152781  0.091615  0.065131     0.151258     0.103391     0.064113   
23  0.190866  0.119009  0.094514     0.053212     0.032421     0.040272   
26  0.135918  0.117934  0.085279     0.017062     0.231259     0.138588   

    offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  ...  aug_vbc_3g  \
1       0.056723      0.092260      0.028195       0.000000  ...    0.000000   
2       0.002031      0.006797      0.003215       0.008728  ...    0.040694   
21      0.001799      0.004487      0.001984       0.000000  ...    0.000000   
23      0.024393      0.048810      0.029353       0.012276  ...    0.000000   
26      0.002456      0.017742      0.004589       0.007378  ...    0.014353   

    jul_vbc_3g  jun_vbc_3g  total_data_recharge_amnt_6  \
1     0.000000      0.0000                         0.0   
2     0.082745      0.0319                         0.0   
21    0.000000      0.0000                         0.0   
23    0.000000      0.0000                         0.0   
26    0.000000      0.0000                         0.0   

    total_data_recharge_amnt_7  total_data_recharge_amnt_8  \
1                     0.000000                    0.000000   
2                     0.000000                    0.000000   
21                    0.000000                    0.000000   
23                    0.001816                    0.000000   
26                    0.000000                    0.018315   

    total_recharge_amnt_6  total_recharge_amnt_7  total_recharge_amnt_8  \
1                0.009067               0.008917               0.008442   
2                0.009067               0.017833               0.016884   
21               0.011352               0.008845               0.007362   
23               0.022596               0.019973               0.019417   
26               0.004715               0.014516               0.032468   

    average_amnt_6_7  
1           0.000701  
2           0.008485  
21          0.002600  
23          0.021967  
26          0.001853  

[5 rows x 140 columns]

## Modelling 

1. Logistic Regression using RFE - To get important predictors for Churn probability
2. Logistic Regression - To know Churn probability + PCA
    - Lasso
    - Ridge
3. Tree models
    - XGBoost
    - Random Forest

## 1. Logistic Regression using RFE - To get importnat predictors for Churn probability

In [52]:
# For RFE
logml = sm.GLM(y_resampled, (sm.add_constant(X_resampled)), family=sm.families.Binomial())
logml.fit().summary()

/Users/kanchoud/opt/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:      churn_probability   No. Observations:                27020
Model:                            GLM   Df Residuals:                    26886
Model Family:                Binomial   Df Model:                          133
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -8626.5
Date:                Wed, 14 Sep 2022   Deviance:                       17253.
Time:                        16:14:51   Pearson chi2:                 1.29e+05
No. Iterations:                   100                                         
Covariance Type:            nonrobust                                         
==============================================================================================
                                 coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------------
const                         -3.6976      0.686     -5.392      0.000      -5.042      -2.354
arpu_6                         7.2522      3.790      1.913      0.056      -0.176      14.681
arpu_7                        19.5193      4.414      4.423      0.000      10.869      28.170
arpu_8                        40.8577      4.960      8.237      0.000      31.136      50.580
onnet_mou_6                  -82.0817     35.829     -2.291      0.022    -152.305     -11.858
onnet_mou_7                  -34.4822     31.247     -1.104      0.270     -95.725      26.760
onnet_mou_8                  -29.1114     47.920     -0.607      0.544    -123.034      64.811
offnet_mou_6                 -78.9793     40.261     -1.962      0.050    -157.890      -0.069
offnet_mou_7                 -32.9949     25.004     -1.320      0.187     -82.003      16.013
offnet_mou_8                 -78.2272     61.017     -1.282      0.200    -197.818      41.364
roam_ic_mou_6                  1.2490      1.092      1.144      0.253      -0.892       3.390
roam_ic_mou_7                  0.8425      1.287      0.654      0.513      -1.681       3.366
roam_ic_mou_8                 -1.5244      1.345     -1.133      0.257      -4.161       1.112
roam_og_mou_6                 39.6031     18.312      2.163      0.031       3.711      75.495
roam_og_mou_7                 16.5520     10.554      1.568      0.117      -4.132      37.237
roam_og_mou_8                 29.2606     23.530      1.244      0.214     -16.856      75.378
loc_og_t2t_mou_6           -3.504e+04   2.54e+04     -1.380      0.168   -8.48e+04    1.47e+04
loc_og_t2t_mou_7            -1.63e+04   2.87e+04     -0.568      0.570   -7.26e+04       4e+04
loc_og_t2t_mou_8            2.156e+05   4.42e+04      4.876      0.000    1.29e+05    3.02e+05
loc_og_t2m_mou_6            -2.56e+04   1.85e+04     -1.381      0.167   -6.19e+04    1.07e+04
loc_og_t2m_mou_7           -1.004e+04   1.77e+04     -0.568      0.570   -4.47e+04    2.46e+04
loc_og_t2m_mou_8            9.952e+04   2.04e+04      4.877      0.000    5.95e+04     1.4e+05
loc_og_t2f_mou_6           -3370.7801   2438.629     -1.382      0.167   -8150.406    1408.846
loc_og_t2f_mou_7           -1794.2490   3162.894     -0.567      0.571   -7993.407    4404.909
loc_og_t2f_mou_8             1.18e+04   2419.397      4.878      0.000    7060.220    1.65e+04
loc_og_t2c_mou_6              -8.9534      1.520     -5.892      0.000     -11.932      -5.975
loc_og_t2c_mou_7              -2.3237      2.551     -0.911      0.362      -7.324       2.677
loc_og_t2c_mou_8               7.1924      1.592      4.519      0.000       4.073      10.312
loc_og_mou_6                3.875e+04   4.76e+04      0.814      0.416   -5.45e+04    1.32e+05
loc_og_mou_7                 -63.2630   3.5

## Lower the p-value higher the significance

In [53]:
logreg = LogisticRegression()

In [54]:
# Running rfe with 70 variables as output
rfe = RFE(logreg, n_features_to_select=70)
rfe = rfe.fit(X_resampled, y_resampled)

/Users/kanchoud/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/kanchoud/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html

In [55]:
rfe.support_ # it givs whether or not the selected feature was in top 70. True means yes

array([ True,  True,  True, False,  True, False, False,  True,  True,
       False, False, False,  True,  True,  True, False, False,  True,
       False,  True,  True, False, False,  True,  True, False, False,
       False, False,  True, False, False, False, False, False,  True,
       False, False,  True,  True,  True,  True, False, False, False,
       False, False,  True, False, False, False, False, False,  True,
       False, False,  True,  True, False,  True, False, False,  True,
        True,  True,  True,  True, False,  True, False, False,  True,
       False, False,  True,  True, False,  True, False, False,  True,
       False, False,  True, False, False,  True, False, False, False,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False,  True, False, False,  True,  True,  True,  True,
        True, False,  True,  True,  True,  True, False, False,  True,
       False,  True,  True,  True, False,  True,  True,  True,  True,
        True, False,

In [56]:
# to see the selected columns
list(zip(X_resampled.columns, rfe.support_, rfe.ranking_))

[('arpu_6', True, 1),
 ('arpu_7', True, 1),
 ('arpu_8', True, 1),
 ('onnet_mou_6', False, 15),
 ('onnet_mou_7', True, 1),
 ('onnet_mou_8', False, 53),
 ('offnet_mou_6', False, 14),
 ('offnet_mou_7', True, 1),
 ('offnet_mou_8', True, 1),
 ('roam_ic_mou_6', False, 5),
 ('roam_ic_mou_7', False, 67),
 ('roam_ic_mou_8', False, 62),
 ('roam_og_mou_6', True, 1),
 ('roam_og_mou_7', True, 1),
 ('roam_og_mou_8', True, 1),
 ('loc_og_t2t_mou_6', False, 46),
 ('loc_og_t2t_mou_7', False, 50),
 ('loc_og_t2t_mou_8', True, 1),
 ('loc_og_t2m_mou_6', False, 13),
 ('loc_og_t2m_mou_7', True, 1),
 ('loc_og_t2m_mou_8', True, 1),
 ('loc_og_t2f_mou_6', False, 18),
 ('loc_og_t2f_mou_7', False, 55),
 ('loc_og_t2f_mou_8', True, 1),
 ('loc_og_t2c_mou_6', True, 1),
 ('loc_og_t2c_mou_7', False, 8),
 ('loc_og_t2c_mou_8', False, 7),
 ('loc_og_mou_6', False, 69),
 ('loc_og_mou_7', False, 24),
 ('loc_og_mou_8', True, 1),
 ('std_og_t2t_mou_6', False, 31),
 ('std_og_t2t_mou_7', False, 11),
 ('std_og_t2t_mou_8', False, 56)

In [57]:
col = X_resampled.columns[rfe.support_]

In [58]:
X_resampled_sm = sm.add_constant(X_resampled[col])
logm2 = sm.GLM(y_resampled, X_resampled_sm, family=sm.families.Binomial())
res = logm2.fit()
res.summary()

/Users/kanchoud/opt/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:      churn_probability   No. Observations:                27020
Model:                            GLM   Df Residuals:                    26950
Model Family:                Binomial   Df Model:                           69
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -8882.4
Date:                Wed, 14 Sep 2022   Deviance:                       17765.
Time:                        16:15:17   Pearson chi2:                 2.09e+05
No. Iterations:                     7                                         
Covariance Type:            nonrobust                                         
==============================================================================================
                                 coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------------
const                         -3.3103      0.624     -5.306      0.000      -4.533      -2.088
arpu_6                         6.6189      3.403      1.945      0.052      -0.052      13.290
arpu_7                        17.7276      3.929      4.512      0.000      10.026      25.429
arpu_8                        37.4215      4.506      8.305      0.000      28.590      46.253
onnet_mou_7                    7.5059      2.133      3.518      0.000       3.324      11.687
offnet_mou_7                   6.5199      1.884      3.461      0.001       2.828      10.212
offnet_mou_8                 -42.8867      9.613     -4.461      0.000     -61.729     -24.045
roam_og_mou_6                  1.3021      0.820      1.587      0.113      -0.306       2.910
roam_og_mou_7                  1.9579      1.078      1.817      0.069      -0.154       4.070
roam_og_mou_8                 16.6329      2.851      5.834      0.000      11.045      22.221
loc_og_t2t_mou_8            1.519e+04   3.85e+04      0.395      0.693   -6.02e+04    9.06e+04
loc_og_t2m_mou_7              -6.1339      1.644     -3.730      0.000      -9.357      -2.911
loc_og_t2m_mou_8            7026.7003   1.77e+04      0.396      0.692   -2.78e+04    4.18e+04
loc_og_t2f_mou_8             832.3677   2104.227      0.396      0.692   -3291.841    4956.576
loc_og_t2c_mou_6              -7.1135      1.253     -5.678      0.000      -9.569      -4.658
loc_og_mou_8               -1.559e+04   3.95e+04     -0.395      0.693    -9.3e+04    6.18e+04
std_og_t2m_mou_8              36.9920      9.651      3.833      0.000      18.076      55.908
std_og_t2f_mou_8             -10.2806      2.484     -4.139      0.000     -15.149      -5.412
std_og_mou_6                   1.6685      0.543      3.071      0.002       0.604       2.734
std_og_mou_7                  -3.2747      2.218     -1.476      0.140      -7.622       1.072
std_og_mou_8                  29.8171     29.560      1.009      0.313     -28.120      87.755
spl_og_mou_8                   1.3218      1.933      0.684      0.494      -2.467       5.111
total_og_mou_8               -38.3608     29.783     -1.288      0.198     -96.734      20.013
loc_ic_t2t_mou_8            2690.3369   8962.952      0.300      0.764   -1.49e+04    2.03e+04
loc_ic_t2m_mou_6               1.3683      1.841      0.743      0.457      -2.240       4.976
loc_ic_t2m_mou_8            3153.0389   1.05e+04      0.301      0.763   -1.74e+04    2.37e+04
loc_ic_t2f_mou_8             983.0354   3293.988      0.298      0.765   -5473.063    7439.133
loc_ic_mou_6                   0.3828      2.489      0.154      0.878      -4.495       5.260
loc_ic_mou_7                  16.8079      1.637     10.269      0.000      13.600      20.016
loc_ic_mou_8               -3637.7143    1.

### Making predictions on the test set

In [59]:
X_test = X_test[col]
X_test.head()

arpu_6    arpu_7    arpu_8  onnet_mou_7  offnet_mou_7  offnet_mou_8  \
5662    554.740   782.352   673.692        38.63        845.83        661.89   
37574  1563.157  1579.675  1256.565      3158.18       1144.44       1327.03   
58341   594.217   796.737   996.393      1200.53        186.48        585.36   
23282   352.160   197.245   365.564       179.88        116.58        507.64   
30561   332.040   275.976   267.033        96.89        178.24        424.68   

       roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_8  ...  \
5662             0.0           0.00            0.0             15.56  ...   
37574            0.0           0.00            0.0              5.91  ...   
58341            0.0           0.00            0.0              8.53  ...   
23282            0.0          34.96            0.0            227.84  ...   
30561            0.0           0.00            0.0              3.78  ...   

       sachet_2g_6  sachet_2g_8  monthly_3g_6  monthly_3g_7  monthly_3g_8  \
5662             0            0             0             0             0   
37574            0            0             0             0             0   
58341            0            0             0             0             0   
23282            0            3             0             0             0   
30561            0            0             0             0             0   

       sachet_3g_6  sachet_3g_8  aug_vbc_3g  jul_vbc_3g  \
5662             0            0        0.00         0.0   
37574            0            0        0.00         0.0   
58341            0            0        0.00         0.0   
23282            0            0        0.00         0.0   
30561            0            0       31.05       337.7   

       total_data_recharge_amnt_8  
5662                          0.0  
37574                         0.0  
58341                         0.0  
23282                       207.0  
30561                         0.0  

[5 rows x 70 columns]

In [60]:
X_test_sm = sm.add_constant(X_test)

/Users/kanchoud/opt/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [61]:
y_test_pred = res.predict(X_test_sm)

/Users/kanchoud/opt/anaconda3/lib/python3.9/site-packages/statsmodels/genmod/families/links.py:188: RuntimeWarning: overflow encountered in exp
  t = np.exp(-z)


In [62]:
y_test_pred[:10]

5662     0.0
37574    1.0
58341    0.0
23282    0.0
30561    1.0
45844    0.0
30159    0.0
62778    0.0
45559    0.0
35274    0.0
dtype: float64

In [63]:
X_test_sm.head()

const    arpu_6    arpu_7    arpu_8  onnet_mou_7  offnet_mou_7  \
5662     1.0   554.740   782.352   673.692        38.63        845.83   
37574    1.0  1563.157  1579.675  1256.565      3158.18       1144.44   
58341    1.0   594.217   796.737   996.393      1200.53        186.48   
23282    1.0   352.160   197.245   365.564       179.88        116.58   
30561    1.0   332.040   275.976   267.033        96.89        178.24   

       offnet_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  ...  \
5662         661.89            0.0           0.00            0.0  ...   
37574       1327.03            0.0           0.00            0.0  ...   
58341        585.36            0.0           0.00            0.0  ...   
23282        507.64            0.0          34.96            0.0  ...   
30561        424.68            0.0           0.00            0.0  ...   

       sachet_2g_6  sachet_2g_8  monthly_3g_6  monthly_3g_7  monthly_3g_8  \
5662             0            0             0             0             0   
37574            0            0             0             0             0   
58341            0            0             0             0             0   
23282            0            3             0             0             0   
30561            0            0             0             0             0   

       sachet_3g_6  sachet_3g_8  aug_vbc_3g  jul_vbc_3g  \
5662             0            0        0.00         0.0   
37574            0            0        0.00         0.0   
58341            0            0        0.00         0.0   
23282            0            0        0.00         0.0   
30561            0            0       31.05       337.7   

       total_data_recharge_amnt_8  
5662                          0.0  
37574                         0.0  
58341                         0.0  
23282                       207.0  
30561                         0.0  

[5 rows x 71 columns]

In [64]:
X_resampled_sm.shape

(27020, 71)

In [65]:
y_pred_train = res.predict(X_resampled_sm)

In [66]:
metric = []
r2_train_lr = r2_score(y_resampled, y_pred_train)
print(f"r-squared train: {r2_train_lr}")
metric.append(r2_train_lr)

rss1_lr = np.sum(np.square(y_resampled - y_pred_train))
print(f"rss train: {rss1_lr}")
metric.append(rss1_lr)

mse_train_lr = mean_squared_error(y_resampled, y_pred_train)
print(f"mse train: {mse_train_lr}")
metric.append(mse_train_lr**0.5)

r-squared train: 0.6148878388269438
rss train: 2601.432648723995
mse train: 0.09627804029326406


## Looking at the rmse value (close to 0.1) Logistic regression doesn't seems to be a good model with such large number of  columns.

## So lets move on with PCA and Logistic Regression to check if model improves or not.

## PCA on the data

- While computing the principal components, we must not include the entire dataset. Model building is all about doing well on the data we haven't seen yet!
- So we'll calculate the PCs using the train data, and apply them later on the test data.